pkg/store/proxy.go

```diff
@@ -50,13 +50,127 @@ type Client interface {
 type ProxyStore struct {
 	logger         log.Logger
 	stores         func() []Client
+	rules          func() []storepb.RulesClient
 	component      component.StoreAPI
 	selectorLabels labels.Labels
 
 	responseTimeout time.Duration
 	metrics         *proxyStoreMetrics
 }
 
+func (s *ProxyStore) Rules(req *storepb.RulesRequest, srv storepb.Rules_RulesServer) error {
+	var (
+		g, gctx  = errgroup.WithContext(srv.Context())
+		respChan = make(chan *storepb.RuleGroup, 10)
+		groups   []*storepb.RuleGroup
+	)
+
+	defer func() { close(respChan) }()
+
+	for _, rulesClient := range s.rules() {
+		rs := &rulesStream{
+			client:  rulesClient,
+			request: req,
+			channel: respChan,
+			server:  srv,
+		}
+		g.Go(func() error { return rs.receive(gctx) })
+	}
+
+	for resp := range respChan {
+		groups = append(groups, resp)
+	}
+
+	if err := g.Wait(); err != nil {
+		level.Error(s.logger).Log("err", err)
+		return err
+	}
+
+	groups = dedupGroups(groups)
+
+	for _, g := range groups {
+		if err := srv.Send(storepb.NewRuleGroupRulesResponse(g)); err != nil {
+			return status.Error(codes.Unknown, errors.Wrap(err, "send rules response").Error())
+		}
+	}
+
+	return nil
+}
+
+func dedupGroups(groups []*storepb.RuleGroup) []*storepb.RuleGroup {
```

Comment:
I thought we are going to do streaming dedup, not convert everything into slices and then dedup?

Reply:
I would love to leave real stream based deduplication as an exercise for the future, as currently I see the problem, that we cannot really stream back to the client any rule, until we received the youngest rule and/or (firing) alert from all replicas.

Reply:
ok future sounds good :+1: